aea/protocols/dialogue/base.py

```diff
@@ -948,6 +995,220 @@ def add_dialogue_endstate(
             self._other_initiated[end_state] += 1
 
 
+def find_caller_object(object_type: Type):
+    """Find caller object of certain type in the call stack."""
+    caller_object = None
+    for frame_info in inspect.stack():
+        frame_self = frame_info.frame.f_locals.get("self", None)
+        if not frame_self:
+            continue
+
+        if not isinstance(frame_self, object_type):
+            continue
+        caller_object = frame_self
+    return caller_object
+
+
+class BasicDialoguesStorage:
+    """Dialogues state storage."""
+
+    def __init__(self, dialogues: "Dialogues") -> None:
+        """Init dialogues storage."""
+        self._dialogues_by_dialogue_label = {}  # type: Dict[DialogueLabel, Dialogue]
+        self._dialogue_by_address = defaultdict(
+            list
+        )  # type: Dict[Address, List[Dialogue]]
+        self._incomplete_to_complete_dialogue_labels = (
+            {}
+        )  # type: Dict[DialogueLabel, DialogueLabel]
+        self._dialogues = dialogues
+
+    def setup(self) -> None:
+        """Set up dialogue storage."""
+
+    def teardown(self) -> None:
+        """Tear down dialogue storage."""
+
+    def add(self, dialogue: Dialogue) -> None:
+        """
+        Add dialogue to storage.
+
+        :param dialogue: dialogue to add.
+        :return: None
+        """
+        self._dialogues_by_dialogue_label[dialogue.dialogue_label] = dialogue
+        self._dialogue_by_address[
+            dialogue.dialogue_label.dialogue_opponent_addr
+        ].append(dialogue)
```

Comment:
Would be cheaper to change this to `_dialogue_label_by_address` and then use the `_dialogues_by_dialogue_label` to retrieve the dialogues! Means, we only store dialogues once

Reply:
actually, it does not copy a dialogue object, it's just a reference. it does not dumped, and can be restored easily from dialogues.

also, this structure used in a couple of cases only, probably tac? so logic overview can be helpful.

Reply:
This functionality is essential even if currently it is just used in tac.